extensions/common/gcp/gcp-core/src/test/java/org/eclipse/edc/gcp/bigquery/BigQueryServiceImplTest.java

```diff
@@ -0,0 +1,364 @@
+            when(streamWriter.append(any())).thenReturn(future);
+        } else {
+            when(streamWriter.append(any())).thenThrow(new IOException());
+        }
+        when(streamWriter.isClosed()).thenReturn(false);
+        when(streamWriter.getStreamName()).thenReturn("name");
+        when(writeClient.finalizeWriteStream(streamWriter.getStreamName())).thenReturn(
+                FinalizeWriteStreamResponse.newBuilder().setRowCount(partsCount).build());
+
+        var commitResponse = mock(BatchCommitWriteStreamsResponse.class);
+        when(commitResponse.hasCommitTime()).thenReturn(true);
+        when(commitResponse.getStreamErrorsList()).thenReturn(Arrays.asList());
+        when(writeClient.batchCommitWriteStreams(any(BatchCommitWriteStreamsRequest.class))).thenReturn(commitResponse);
+
+        var executorService = Executors.newSingleThreadScheduledExecutor();
+        executorService.schedule(() -> bigQueryService.testAppendSignal(), 2, TimeUnit.SECONDS);
+
+        bigQueryService.runSinkQuery(parts);
+
+        verify(streamWriter, times(parts.size())).append(any());
+        verify(streamWriter).close();
+        verify(writeClient).finalizeWriteStream("name");
+        if (succeeds) {
+            verify(writeClient).batchCommitWriteStreams(any(BatchCommitWriteStreamsRequest.class));
+        } else {
+            // If an error occurred while appending rows, then data should not be committed.
+            verify(writeClient, never()).batchCommitWriteStreams(any(BatchCommitWriteStreamsRequest.class));
+        }
+    }
+
+    @Test
+    void testRunSinkQuerySucceeds() throws DescriptorValidationException, IOException, InterruptedException, ExecutionException, TimeoutException {
+        testRunSinkQuery(true);
+    }
+
+    @Test
+    void testRunSinkQueryFailsNoCommit() throws DescriptorValidationException, IOException, InterruptedException, ExecutionException, TimeoutException {
+        testRunSinkQuery(false);
+    }
+
+    private void testRunSourceQuery(String sinkType) throws InterruptedException, IOException {
+        var sinkAddress = DataAddress.Builder.newInstance()
+                .type(sinkType)
+                .property("customerName", TEST_CUSTOMER_NAME)
+                .build();
+
+        var queryJob = mock(Job.class);
+        var tableResult = mock(TableResult.class);
+        var schema = mock(Schema.class);
+        var jobStatus = mock(JobStatus.class);
+
+        when(bigQuery.create(any(JobInfo.class))).thenReturn(queryJob);
+        when(queryJob.getQueryResults(QueryResultsOption.pageSize(4))).thenReturn(tableResult);
+        when(tableResult.getSchema()).thenReturn(schema);
+        when(tableResult.hasNextPage()).thenReturn(false);
+        when(schema.getFields()).thenReturn(FieldList.of(TEST_FIELDS));
+        when(tableResult.getValues()).thenReturn(Arrays.asList(
+                FieldValueList.of(TEST_ROW_1),
+                FieldValueList.of(TEST_ROW_2)
+        ));
```

Comment:
a lot of mocks here, looks like this piece of code would likely need an higher level test that uses some real components

Reply:
Reduced mocks